service/header/sync_ranges.go

```diff
@@ -0,0 +1,148 @@
+package header
+
+import "sync"
+
+// ranges keeps non-overlapping and non-adjacent header ranges which are used to cache headers (in ascending order).
+// This prevents unnecessary / duplicate network requests for additional headers during sync.
+type ranges struct {
+	ranges []*_range
+	lk     sync.Mutex // no need for RWMutex as there is only one reader
+}
+
+// Head returns the highest ExtendedHeader in all ranges if any.
+func (rs *ranges) Head() *ExtendedHeader {
+	rs.lk.Lock()
+	defer rs.lk.Unlock()
+
+	ln := len(rs.ranges)
+	if ln == 0 {
+		return nil
+	}
+
+	head := rs.ranges[ln-1]
+	return head.Head()
+}
+
+// Add appends the new ExtendedHeader to existing range or starts a new one.
+// It starts a new one if the new ExtendedHeader is not adjacent to any of existing ranges.
+func (rs *ranges) Add(h *ExtendedHeader) {
+	head := rs.Head()
+
+	// short-circuit if header is from the past
+	if head != nil && head.Height >= h.Height {
+		// TODO(@Wondertan): Technically, we can still apply the header:
+		//  * Headers here are verified, so we can trust them
+		//  * PubSub does not guarantee the ordering of msgs
+		//    * So there might be a case where ordering is broken
+		//    * Even considering the delay(block time) with which new headers are generated
+		//    * But rarely
+		//  Would be still nice to implement
+		log.Warnf("rcvd headers in wrong order")
```

Comment:
```suggestion
		log.Warnf("received headers in wrong order")
```
Maybe even add, where the current head is and what was received. 

Orthogonal question: Is it possible to filter the logs per module/package? (e.g. if I want to only see the header related warnings etc).

Reply:
> is it possible to filter the logs per module/package?

Yes